### ISSUE TITLE
Making use of releases

### ISSUE THREAD

**Issue by dimsanius** [Meru3m/virtualbox-snapshotter#9]
I think it may be a good idea to utilise releases.

Repo contains lots of readmes and images now an is already at `1.02MB`. That is just the size of the files.
However, GitHub clones the history of the files too resulting in total size cloned of `2.35MB`
It will grow even further as time goes by.
This may impose issues for people using metered networks.

Therefore, is it possible to introduce releases? 
If yes, I think it may be good to get releases going from very beginning.
Release 1.0.0 - commit id 5eee828f31f17cd115c3f30b5e6c92e38c37954e
Release 1.0.1 - commit id 6c5d2c6e176dee246e19863da87568cd4821589e
Release 1.1.0 - commit id 35e23050ce79668082cc334b5d41803af4fd5a2e

PR https://github.com/Meru3m/virtualbox-snapshotter/pull/8 does not affect the script aside from adding a comment in. So, it is not sufficient to make a release after that specific PR is/if/when merged .

Let me know if any help needed with changelog.

**Comment by dimsanius** [comment]
By default, GitHub just zips the entire repo. This is not a desired behaviour. However, this can be adjusted

It is required to add `.gitattributes` file and ignore all irrelevant files for release within it. 
Source: https://stackoverflow.com/a/58443948

`.gitattributes` file should look something like this:
```
README.md export-ignore
.gitignore export-ignore
...
```

It looks simple enough. I tested it on my local fork and it works well. Let me know if you like that - I will do a PR with `.gitattributes` if you agree ^_^

**Comment by dimsanius** [comment]
Since this project is simple enough, only 2 files need to reside within release:
- Main python script file
- Requirements file

However, since I introduced `prospector` to lint the code, it may be a good idea to have 2 requirements files:
- `requirements.txt` - a file without a `prospector`. This will go into each `release`
- `requirements_dev.txt` - a file with a `prospector` to allow linting when repo cloned using `git clone`
Prospector installs tons of other libraries which again - may be a trouble for users using metered networks. Thats how I see it.

**Comment by mattiabaldari** [comment]
Hi @dimsanius,

first, regarding versioning, I definitely agree.
Usually I use tags to do so, do you use other methods? If not, I can proceed with tagging the commits you suggested.

second, regarding `.gitattributes` do you know if there's a way an user can specify if he wants to download everything or just the "reduced" repo? This way someone can choose if he also wants the guide and all the rest or just the light version. In general I think it is a good idea and it helps to keep the repository not so big. [😉](https://emojipedia.org/it/faccina-che-fa-l-occhiolino/)

third, couldn't we work on two different branch? `develop` and `master` and we keep two different `requirements.txt` on each of them. Let me know what you think.

Thank you 🙏

**Comment by dimsanius** [comment]
TL;DR

1. Tagging is the way to go
2. `requirements.txt` and main python file are the only files to go into a release
3. Need some more research and thinking put into handling `prospector` in `requirements.txt`

First of all, regarding versioning - we can go both ways - GitHub allows to specify a commit OR a tag for a release. I prefer latter - tags are much more useful as they will explicitly say which exact commit hash the release was made under. It will allow to select exact source code for exact version from the root of repo as well as release page. Whereas, commit ids will not be marked in any way within root of a repo. Therefore, I would say that we should employ tagging as it is a set in stone practice.

Secondly, regarding `.gitattributes` file - ignoring files for release via '.gitattributes' will not impact source code file size. All the files will still be stored within repo. However, it is be possible to ignore any files we don't want to be added to a release. This will result in a size reduction of a release. 
Adding all files to a release is kind of redundant because cloning the repo will yield the same result. This will negate the point of releases. I would say we should go for only the `requirements.txt` and main python file for releases.
Keep in mind that releases allow users to streamline their own update mechanism as updates to `virtualbox-snapshotter` are implemented.

Thirdly, regarding `master` and `develop` branches - I like the approach and that is how it should be done. However, this may not be the case with this repo. Usual practice is that `develop` is used for development and may contain unstable code while `master` is up-to-date, tested and fully working code. Therefore, `develop` has to be merged into `master` when code is stable enough and has enough new features implemented. However, if its merged, all contents will be merged including `requirements.txt` file. And we do not want that. Moreover, proposed `master` and `develop` approach implies that both branches do not rely on each other, meaning that when a new change is made, it has to be PR'd into both branches. This is quite hard to impose in case we will have any new contributors. So, we may think of another approach. 
So far the options I can think of:

- Have 2 requirements files: `requirements_dev.txt` for development with `prospector` (`requirements_dev.txt` can be ignored via `.gitattributes` when preparing a release), `requirements.txt` for release
- Employ some kind of automation to handle `requirements.txt` for us when creating a release
- Leave releases as is with the `prospector` module, resulting in extra data to be downloaded upon install
- Discard `prospector` all together (I really do not like this approach - linting is good, especially when it works. And it works quite well with this repo)

Let me know what you think. Let's not rush and do some reading first. So, I will sleep on it and may come back with more info.

**Comment by dimsanius** [comment]
I have just figured out - if no releases are created GitHub uses tags to display under the `Releases`. I think we can utilise that. Seems like some bigger projects use just the tags for releases instead of using the actual full-fledged release: 
[https://github.com/jaxwilko/gtk-theme-framework](https://github.com/jaxwilko/gtk-theme-framework)
[https://github.com/robertalexa/MMM-OpenWeatherMapForecast](https://github.com/robertalexa/MMM-OpenWeatherMapForecast)

And by the looks of it, it does not zip the entire repo with all the commit history and history of every file. It just zips the contents of the repo at that specific tag. This means that all the hassle of creating a full fledged release can be discarded until someone actually requests such a feature. And `.gitattributes` rules are still being applied when zip archives are created at tags.

 I think we can stick with that. I figured out that full-fledged releases are only used when there are some files that need to be built. I.e when compiling C code into `bin` file or something: [https://github.com/Aircoookie/WLED](https://github.com/Aircoookie/WLED). And since this repo is based on Python, there is no need to create a full fledged release because there is no files to be compiled. Just tag will be sufficient.

However, this still does not sort the issue with `prospector`

**Comment by dimsanius** [comment]
I propose following:

1. We push a `.gitattributes` file with all unnecessary files ignored - this will make only the latest (and every upcoming) release clean.
2. Tag previous and current versions as described in the first message - old versions will have entire repo zipped. Unfortunately, we cannot go back in time to fix that :c

Please ensure no other tags are pushed rather then semantic version tags `vX.Y.Z` i.e. `v1.0.0`, `v1.0.1`, `v1.1.0` etc.

However, `prospector` question is still up for debate... Let me know your ideas.

**Comment by mattiabaldari** [comment]
Hi @dimsanius,

1) I tagged the repo and created the tag protection rule.
2) Let's try to use `.gitattributes` and let's see how it is. In case we like the approach we can introduce it in the next release.
3) In order not to introduce complexity for now let's just have `requirements_dev.txt` and `requirements.txt`. The idea to use `develop` and `master` branch it is because I was thinking to use gitflow, but maybe it is an overkill. For now we can just keep the two requirements.

Do you agree?

**Comment by dimsanius** [comment]
Heyo,

1. Great! I fully support this idea.
2. Sure, let me do a PR. Shouldn't take me long
3. Yea, 2 requirements makes it a bit more straight forward

I have noticed you have done the releases. I will add the changelogs.

**Comment by dimsanius** [comment]
Please review my PR https://github.com/Meru3m/virtualbox-snapshotter/pull/10. I did a woopsy and pushed to the master branch of my forked repo :/

**Comment by dimsanius** [comment]
I have updated release descriptions with changelogs

**Comment by mattiabaldari** [comment]
Thank you, I commented it.
